storage/database/sharded_database.go

```diff
@@ -151,102 +154,283 @@ func (pdb *shardedDB) Close() {
-func (pdb *shardedDB) NewIteratorWithStart(start []byte) Iterator {
-	// TODO-Klaytn implement this later.
-	iterators := make([]Iterator, 0, pdb.numShards)
-	for i := 0; i < int(pdb.numShards); i++ {
-		iterators = append(iterators, pdb.shards[i].NewIteratorWithStart(start))
+func (db *shardedDB) NewIteratorWithStart(start []byte) Iterator {
+	return db.newIterator(func(db Database) Iterator { return db.NewIteratorWithStart(start) })
+}
+
+// NewIteratorWithPrefix creates a binary-alphabetical iterator over a subset
+// of database content with a particular key prefix.
+func (db *shardedDB) NewIteratorWithPrefix(prefix []byte) Iterator {
+	return db.newIterator(func(db Database) Iterator { return db.NewIteratorWithPrefix(prefix) })
+}
+
+func (db *shardedDB) newIterator(newIterator func(Database) Iterator) Iterator {
+	it := &shardedDBIterator{
+		db.NewChanIterator(context.Background(), newIterator),
+		make(chan entry, shardedDBCombineChanSize),
+		nil, nil}
+
+	go it.newCombineWorker()
+
+	return it
+}
+
+// newCombineWorker fetches any key/value from resultChs and put the data in resultCh
+// in binary-alphabetical order.
+func (it shardedDBIterator) newCombineWorker() {
+	entries := make([]entry, len(it.resultChs)) // contains smallest values from each iterators
+
+	// fill initial values for entries
+	for i := len(it.resultChs) - 1; i >= 0; i-- {
+		ch := it.resultChs[i]
+		e, ok := <-ch
+		if !ok { // channel is closed
+			it.resultChs = append(it.resultChs[:i], it.resultChs[i+1:]...)
+			entries = append(entries[:i], entries[i+1:]...)
+		} else {
+			entries[i] = e
+		}
 	}
 
-	for _, iter := range iterators {
-		if iter != nil {
-			if !iter.Next() {
-				iter = nil
+chanIter:
+	for len(it.resultChs) != 0 {
+		// check if done
+		select {
+		case <-it.ctx.Done():
+			logger.Trace("[shardedDBIterator] combine worker ends due to ctx")
+			break chanIter
+		default:
+		}
+
+		// look for smallest key
+		minEntry, minIdx := entries[0], 0
+		for i, e := range entries[1:] {
+			if bytes.Compare(minEntry.key, e.key) > 0 {
+				minEntry = e
+				minIdx = i + 1
 			}
 		}
+
+		// fill resultCh with smallest key
+		it.resultCh <- minEntry
+
+		// fill used entry with new entry
+		e, ok := <-it.resultChs[minIdx]
+		if !ok { // channel is closed
+			it.resultChs = append(it.resultChs[:minIdx], it.resultChs[minIdx+1:]...)
+			entries = append(entries[:minIdx], entries[minIdx+1:]...)
+		} else {
+			entries[minIdx] = e
+		}
 	}
+	logger.Trace("[shardedDBIterator] combine worker finishes")
+	close(it.resultCh)
+}
 
-	return &shardedDBIterator{iterators, nil, nil}
+// Next gets the next item from iterators.
+// The first call of Next() might take 50ms.
+func (it *shardedDBIterator) Next() bool {
+	maxRetry := 5
+	for i := 0; i < maxRetry; i++ {
+		select {
+		case e, ok := <-it.resultCh:
+			if ok {
+				it.key, it.value = e.key, e.val
+				return true
+			} else {
+				logger.Error("[shardedDBIterator] Next is called on closed channel")
```

Comment:
It seems that this log is always printed when the channel is closed. 
But `close(it.resultCh)` is not error case.